File: kirun-js/src/engine/function/system/array/Copy.ts
import { KIRuntimeException } from '../../../exception/KIRuntimeException';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { ArrayUtil } from '../../../util/ArrayUtil';
import { MapUtil } from '../../../util/MapUtil';
import { isNullValue } from '../../../util/NullCheck';
import { StringFormatter } from '../../../util/string/StringFormatter';
import { AbstractArrayFunction } from './AbstractArrayFunction';

export class Copy extends AbstractArrayFunction {
    public constructor() {
        super(
            'Copy',
            [
                Copy.PARAMETER_ARRAY_SOURCE,
                Copy.PARAMETER_INT_SOURCE_FROM,
                Copy.PARAMETER_INT_LENGTH,
                Copy.PARAMETER_BOOLEAN_DEEP_COPY,
            ],
            Copy.EVENT_RESULT_ARRAY,
        );
    }

    protected internalExecute(context: FunctionExecutionParameters): FunctionOutput {
        var source = context.getArguments().get(Copy.PARAMETER_ARRAY_SOURCE.getParameterName());
        var srcfrom = context.getArguments().get(Copy.PARAMETER_INT_SOURCE_FROM.getParameterName());
        var length = context.getArguments().get(Copy.PARAMETER_INT_LENGTH.getParameterName());

        if (length == -1) length = source.length - srcfrom;

        if (srcfrom + length > source.length)
            throw new KIRuntimeException(
                StringFormatter.format(
                    'Array has no elements from $ to $ as the array size is $',
                    srcfrom,
                    srcfrom + length,
                    source.length,
                ),
            );

        var deep = context.getArguments().get(Copy.PARAMETER_BOOLEAN_DEEP_COPY.getParameterName());

        const ja: any[] = new Array(length);

        for (let i = srcfrom; i < srcfrom + length; i++) {
            if (!isNullValue(source[i]))
                ja[i - srcfrom] = deep ? JSON.parse(JSON.stringify(source[i])) : source[i];
        }

        return new FunctionOutput([EventResult.outputOf(MapUtil.of(Copy.EVENT_RESULT_NAME, ja))]);
    }
}

File: kirun-py/src/kirun_py/function/system/array/copy.py
from __future__ import annotations

from typing import Any, List, TYPE_CHECKING

from kirun_py.exception.ki_runtime_exception import KIRuntimeException
from kirun_py.function.system.array.abstract_array_function import AbstractArrayFunction
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.util.duplicate import duplicate
from kirun_py.util.null_check import is_null_value

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class Copy(AbstractArrayFunction):

    def __init__(self) -> None:
        super().__init__(
            'Copy',
            [
                AbstractArrayFunction.PARAMETER_ARRAY_SOURCE,
                AbstractArrayFunction.PARAMETER_INT_SOURCE_FROM,
                AbstractArrayFunction.PARAMETER_INT_LENGTH,
                AbstractArrayFunction.PARAMETER_BOOLEAN_DEEP_COPY,
            ],
            AbstractArrayFunction.EVENT_RESULT_ARRAY,
        )

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        source = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_ARRAY_SOURCE.get_parameter_name()
        )
        srcfrom = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_INT_SOURCE_FROM.get_parameter_name()
        )
        length = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_INT_LENGTH.get_parameter_name()
        )

        if length == -1:
            length = len(source) - srcfrom

        if srcfrom + length > len(source):
            raise KIRuntimeException(
                f'Array has no elements from {srcfrom} to {srcfrom + length} '
                f'as the array size is {len(source)}'
            )

        deep = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_BOOLEAN_DEEP_COPY.get_parameter_name()
        )

        ja: List[Any] = [None] * length

        for i in range(srcfrom, srcfrom + length):
            if not is_null_value(source[i]):
                ja[i - srcfrom] = duplicate(source[i]) if deep else source[i]

        return FunctionOutput([
            EventResult.output_of(
                {AbstractArrayFunction.EVENT_RESULT_NAME: ja}
            )
        ])
